source/mainnet/smart-contracts/onboarding-guide-ethereum-developers/faq.rst

```diff
@@ -0,0 +1,760 @@
+
+    Error codes come from the ``concordium-std`` crate or are thrown by the smart contract itself.
+
+    **Concordium-std crate errors**
+
+    An example of a ``concordium-std`` crate error is shown below:
+
+    .. code-block:: console
+
+        Error: Updating contract instance failed:
+        'smash' in 'PiggyBank' at {"index":12345,"subindex":0} failed with code -2147483636.
+
+    .. image:: ../tutorials/piggy-bank/images/pb_tutorial_21.png
+            :width: 100 %
+
+    You can find the meaning of common error codes in this `list <https://docs.rs/concordium-std/3.0.0/concordium_std/#signalling-errors>`_.
+    This error was thrown by the ``concordium-std`` crate. The error codes start from ``i32::MIN`` and go upwards.
+    ``-2147483636`` corresponds to ``NotPayableError``. In contrast, error types of smart contracts start from ``-1`` and go downwards.
+
+    **Smart contract errors**
+
+    An example of a smart contract error is shown below:
+
+    .. code-block:: console
+
+        Error: Updating contract instance failed:
+        'smash' in 'PiggyBank' at {"index":12345,"subindex":0} failed with code -1.
+
+    .. image:: ../tutorials/piggy-bank/images/pb_tutorial_27.png
+        :width: 100 %
+
+    You can find the meaning of the piggy bank error codes in this :ref:`section<piggy-bank-smash-error>`.
+    This error was thrown by the smart contract. The error codes start from ``-1`` and go downwards.
+    ``-1`` is the first error code and corresponds to ``NotOwner``.
+    In contrast, errors from the ``concordium-std`` crate start from ``i32::MIN`` and go upwards.
+
+.. dropdown:: Is there a smart contract code linter?
+
+    Yes. You can use the `fmt` and the `cargo clippy` linter tools as described in the `README <https://github.com/Concordium/concordium-rust-smart-contracts>`_.
+
+.. dropdown:: Are tests executed in parallel or sequentially?
+
+    Tests run in parallel by default by the following two commands
+
+    .. code-block:: console
+
+        $cargo concordium test
+
+    .. code-block:: console
+
+        $cargo test
+
+    You can use the below command for consecutive execution of the tests.
+
+    .. code-block:: console
+
+        $cargo test -- --test-threads=1
+
+.. dropdown:: How do I embed a schema into a smart contract? Why do I need a schema? How can I provide the input parameters as a JSON object and get the output parameters in a human-readable format when using the `concordium-client`?
+
+    A :ref:`smart contract schema<contract-schema>` is a description of how to represent
+    bytes in a more structured representation. It is used by
+    external tools (such as the ``concordium-client``) when displaying the return value of a receive
+    function and for specifying the input parameters to a receive
+    function using a structured representation, such as JSON.
+
+    That the input parameters, and output parameters of a function are included into a schema
```

Comment:
There is something missing in this paragraph, a connector perhaps. It doesn't read quite right.

Reply:
I updated it to:
    The smart contract function needs to define a ``parameter`` and a ``return_value``
    so that a schema can be generated for the input parameters and the output parameters, respectively.